gnosis_vpn/src/main.rs

```diff
@@ -176,33 +160,36 @@ fn incoming_stream(state: &mut core::Core, res_stream: Result<net::UnixStream, c
             return;
         }
     };
-    tracing::debug!(command = %cmd, "parsed command");
 
-    let res = match state.handle_cmd(&cmd) {
+    tracing::debug!(command = %cmd, "incoming command");
+
+    let resp = match core.handle_cmd(&cmd) {
         Ok(res) => res,
         Err(e) => {
-            // Log the error and its chain in one line
-            let error_chain: Vec<String> = e.chain().map(|cause| cause.to_string()).collect();
-            tracing::error!(?error_chain, "error handling command");
+            tracing::error!(error = ?e, "error handling command");
             return;
         }
     };
 
-    if let Some(resp) = res {
-        tracing::info!(response = %resp);
-        if let Err(e) = stream.write_all(resp.as_bytes()) {
-            tracing::error!(error = ?e, "error writing response");
-            return;
-        }
-        if let Err(e) = stream.flush() {
-            tracing::error!(error = ?e, "error flushing stream");
+    let str_resp = match serde_json::to_string(&resp) {
+        Ok(res) => res,
+        Err(e) => {
+            tracing::error!(error = ?e, "error serializing response");
             return;
         }
+    };
+
+    if let Err(e) = stream.write_all(str_resp.as_bytes()) {
+        tracing::error!(error = ?e, "error writing response");
```

Comment:
This should work more nicely:
```suggestion
        tracing::error!(error = %e, "error writing response");
```